docs/proposals/cli.md

```diff
@@ -0,0 +1,188 @@
+
+```sh
+$ shp <verb> <resource> <name> [options]
+```
+
+For instance:
+
+```sh
+$ shp create build nodejs-ex
+$ shp run build nodejs-ex
+```
+
+Alternatively, we may have a more conventional pattern, as in:
+
+```sh
+$ shp <resource> <verb> <name> [options]
+```
+
+For example:
+
+```sh
+$ shp create build build nodejs-ex
+$ shp run build nodejs-ex
+$ shp logs nodejs-ex
+```
+
+During the review process of this enhancement proposal, we must decide which pattern suits better,
+which one we prefer to go ahead with.
+
+## Features
+
+For the initial release, the aim is to be able to define a `Build`, run it and display logs.
+
+### Defining a Build
+
+```sh
+$ shp <verb> build <name> [options]
+```
+
+Will be responsible for managing Build objects, using the verbs: `create`, `update`, list and
+`delete`. Later on, end users should have the ability to bootstrap a local repository clone using a
+single command.
+
+### Running a Build
+
+```sh
+$ shp run build <name> [options]
+```
+
+This sub-command will instantiate a new BuildRun resource for a Build, effectively starting the
+build process. And, will print follow up commands that may be issued to see the BuildRun status
+and inspect logs.
+
+### Logs
+
+```sh
+$ shp logs <name> [options]
+```
+
+Retrieve all logs related to the informed `BuildRun` name. It will also be able to follow
+(`--follow`) container log output as they are executed. The log lines displayed are organised by
+the sequence of steps, and easy to read the whole build process output in a single go.
+
+Additionally, it must provide easy syntax to reach a specific `BuildRun` instance, or the most
+recent `BuildRun` for a `Build`.
+
+## More Features
+
+The command-line interface is planned to host more features, like for instance managing local and
+remove artifacts, as well as help end-users to upload data into the cluster.
+
+Thus, the initial design of the CLI must allow more subcommands to be added, accommodating more
+features which will depend in a client.
+
+## Testing
+
+The strategy to test the command-line application will be based on `go test` for the unit testing,
+and for end-to-end testing we should adopt [bats](https://github.com/sstephenson/bats). Bats
+framework gives us a structured way to run shell script commands and what we expect them to
+return, the command-line client shp is no more than a shell command.
+
+The testing structure will be composed of:
+
+- **Unit**: written in Golang and using [Gomega](https://onsi.github.io/gomega/) for assertion;
+- **End-to-End**: written in Golang and located at the traditional `test/e2e` directory;
+- **System**: written using Bats framework (Bash);
+
+Bats will also be helpful for a future `shp` container-image, we are able to mount the Bats files
+in the container-image produced, and run our system testing against it. Therefore, Bats covers
+testing of a local command-line binary, as well as it does a container-image.
```

Comment:
Quick question regarding the proposed tools: I have worked with both `bats` and Ginkgo. Definitely good way to do it. However, I never really made friends with `bats`, especially when trying to debug an issue and I was missing some ad-hoc flexibility. For one of our tools, I adapted a style I saw in the open-source tool [spruce](https://github.com/geofffranks/spruce). A [helper function in the test suite works as a stand-in for the binary](https://github.com/homeport/dyff/blob/main/internal/cmd/cmd_suite_test.go#L87-L92) and you keep writing all your tests in Go while the [test definition looks kind of like your command line](https://github.com/homeport/dyff/blob/main/internal/cmd/cmds_test.go#L66) would. Maybe you can have a look at the idea.

Reply:
Taking your experience in consideration I'm rewording to make Bats optional, we may decide for another tool depending on the group experience it. I think it's fair to give it a try, follow [buildah's example](https://github.com/containers/buildah/blob/master/tests/config.bats).

Reply:
The `buildah` example looks nice. Something like that could be fun. Looking forward to see this in action.